File: problems/0001-two-sum/solution.ts
/**
 * Возвращает индексы двух чисел из nums, дающих в сумме target.
 * Один проход с хеш-таблицей: O(n) по времени и памяти.
 */
export function twoSum(nums: number[], target: number): number[] {
  const seen = new Map<number, number>();
  for (let i = 0; i < nums.length; i++) {
    const complement = target - nums[i];
    if (seen.has(complement)) return [seen.get(complement)!, i];
    seen.set(nums[i], i);
  }
  return [];
}

File: problems/0001-two-sum/solution.go
package twosum

// TwoSum возвращает индексы двух чисел из nums, дающих в сумме target.
// Решение за один проход с хеш-таблицей: O(n) по времени и памяти.
func TwoSum(nums []int, target int) []int {
	seen := make(map[int]int, len(nums))
	for i, n := range nums {
		if j, ok := seen[target-n]; ok {
			return []int{j, i}
		}
		seen[n] = i
	}
	return nil
}
